doc/man-source/mode2.1
.TH MODE2 "1" "Last change: Jan 2015" "mode2 @version@" "User Commands"
.SH NAME
mode2 - shows pulse/space length or decoded input from infrared drivers

.SH SYNOPSIS
.B mode2
[\fIoptions\fR]
.SH DESCRIPTION
The main purpose of these programs is to check operation of your
home-brew LIRC receiver hardware and to see the IR waveform of the
remote controller without an expensive oscilloscope. Very useful for
debugging also in situations you just need to verify configuration.

mode2 will simply print pulse and space lengths to stdout for drivers
supporting LIRC_MODE_MODE2. For drivers supporting LIRC_MODE_LIRCCODE such
as TV-cards and the irman driver mode2 writes the decoded, integer values.

.SH OPTIONS

.TP
\fB\-h\fR \fB\-\-help\fR
display usage summary
.TP
\fB\-v\fR \fB\-\-version\fR
display version
.TP
\fB\-d\fR \fB\-\-device\fR=\fIdevice\fR
read from given device
.TP
\fB\-H\fR \fB\-\-driver\fR=\fIdriver\fR
use given driver
.TP
\fB\-m\fR \fB\-\-mode\fR
enable alternative display mode
.TP
\fB\-r\fR \fB\-\-raw\fR
access device directly
.TP
\fB\-A\fR \fB\-\-driver\-options=key:value[|key:value...]\fR
Set driver-specific option named key to given value.


.SH RUNNING AS ROOT
In many cases mode2 needs to run as root to access devices not available
to regular users. On the other hand, running as root creates problems
such as log files owned by root, security concerns etc.
.P
In order to cope with this, mode2 by default drops root privileges
after opening the input device. This support is based on that root
permissions are acquired using sudo(1) e. g., using
.nf

        $ sudo mode2 --raw /dev/lirc0

.fi
If not using sudo, the same behaviour could be accomplished using the
SUDO_USER environment variable e. g.,
.nf

        # SUDO_USER=$LOGNAME mode2 --raw /dev/lirc0

.fi
The --keep-root option will make mode2 to keep root privileges for the
complete run.


.SH "DRIVER LOADING"
Drivers are loaded dynamically. The search path used for this is determined
by (falling priority):
.IP \- 2
The --plugindir option if available.
.IP \- 2
The 'plugindir' entry in  the [lircd] section of the lirc_options.conf file.
.IP \- 2
The environment variable LIRC_PLUGINDIR.
.IP \- 2
A hardcoded default (usually /usr/lib[64]/lirc/plugins).

.SH ENVIRONMENT
.TP 4
.B LIRC_LOGLEVEL
Determines the amount of logging information. [level] can be a symbolic
syslog level: 'error','warning, 'info', 'notice' or  'debug'. lircd
also defines three additional levels 'trace', 'trace1' and 'trace2' which
gives even more messages ('trace2' bringing the most). However, in the
log these messages are marked as 'debug'.
.IP
The level can also be an integer in the range 3 (almost no messages) to
10.


.SH FILES

.TP 4
.B ~/.cache/mode2.log
Debug output. Setting the XDG_CACHE_HOME environment variable relocates this
file to $XDG_CACHE_HOME/mode2.log

.SH "SEE ALSO"
xmode2(1)
.br
lirc_ioctl(4)
.br
lirc(4)

The documentation for
.B lirc
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.2-devel
  * The smode2 program has been retired from the build, although it still
    is available in contrib/.
  * The setup.sh script is removed (build always build all drivers).
  * Drivers are loaded dynamically. Many programs including lircd now
    needs to know where the drivers are located to be able to use
    them. There are new options, environment variables and hard-coded
    defaults. See the manpages for more.
  * A new tool lirc-lsplugins is added to inspect dynamic modules.
  * New, unstable configuration script lirc-setup. The script guides user
    when selecting both capture device and remote config. It sports auto-
    detection of most devices, searching and downloading from the remotes
    database and a considerable amount of bugs.
  * The remotes/ directory has become obsolete, only used for testing.
    New configuration tools uses the remotes database at
    http://sf.net/p/lirc-remotes as the only source (#53). A new tool
    irdb-get(1) supports searching and downloading from remotes database.
  * Building on MacOSX has been refreshed.
  * Support for for sonyir on MacOSX is added. This is a single hiddev device;
    with the framework in place hopefully more devices will be added over time.
  * Support for the Dangerous Prototypes irtoy device is added (#20).
  * Support for the DIY assembly kit yard2 usb ir device is added. This is
    an experimental driver which depends on a server which currently is not
    packaged in any major distro; a Debian/Ubuntu PPA is available.
  * Support for long keypresses has been added. See the "Configuring LIRC"
    chapter in the manual, search for "ignore_first_event" (#50).
  * lircd has been verified to run as regular user as long as it can open
    devices and such. A new option --effective-user makes lircd drop it's
    privileges after opening devices if started as root. A new udev rule
    is shipped to make /dev/lirc* accessible for users in group 'lirc" (#56).
  * Logging runtime option is renamed: debug -> loglevel. Levels can be given
    symbolically e. g., 'info' or 'error'. Old option name is still supported.
  * The default location for the user lircrc-file has been changed from
    ~/.lircrc to ~/.config/lircrc. The old location is still used as fallback.
    (ticket #52).
  * The lircd.conf configuration has been updated. lircd will now besides
    the lircd.conf file also read all *.conf files in the
    /etc/lirc/lircd.conf.d directory. The already existing include directive
    has been modified to accept wildcards and also been documented.
  * lircd.conf has a new 'driver' attribute for configs requiring a specific
    driver. Supported by irrecord(1), used by lirc-setup(1) and irdb-get(1).
  * Mechanisms to disable the implicit sorting of remotes has been added,
    and the automatic sorting has been enhanced (ticket #56).
  * The enabling of lircrcd mode synchronization daemon has been changed. The
    old "#! lircrcd" shebang is deprecated and a new, global lircrc option
    "lircrc_class" is added (#62). The lircrcd sockets has been moved to
    /var/run/lirc (#61). This might break some corner-case installations.

  * Dependencies:
      + Build scripts now requires python3.
      + Build requires the pkg-config package, in particular
        /usr/share/aclocal/pkg.m4
      + python3-PyYAML is strongly recommended. lirc builds without it, but
        crippled (no table.html or lirc.hwdb, lirc-setup doesn't work).
      + Doxygen is required to build api docs (ignored if missing).
      + man2html is required to generate html manpages (the bundled version
        is dropped). If missing, no html manpages are generated.
      + lirc-setup requires gtk3 libraries and python3-gobject bindings,
        these are not required to build.
      + help2man is no longer required.
      + kernel sources are not required.
  * New configure options:
      --with-systemdsystemunitdir=... Installation path for systemd files
        (ticket #49).
      --with_lockdir=... Serial lock directory, /var/lock or /var/lockdev
        (ticket #44).
      --enable-python3-fix: Install python3 wrapper to actual python3.x
        version, bug walk-around for MacOSX.
  * Several options to configure have been dropped:
      --with-syslog: Logging defaults to using syslog, but using
        the --logfile runtime option switches to file loggging.
      --enable-debug: All builds have debug enabled.
      --daemonize: Always enabled.
      --enable-sandboxed: Devices are created in runtime.
      --with-driver=... The build always builds all drivers
      --enable-maintainer-mode: lircd.simsend and lircd.simrec
        (test tools) are always created although not installed.
  * The lircd and lircmd sockets are not created at installation any more,
    they are created on demand by the servers. If there already is e. g., a
    /var/run/lirc/lircd socket it will still be used without changing
    permissions (#46).
  * The long-standing issues with parallel make are fixed.
  * Many more files are installed in the data directory e. g., configuration
    files and a local copy of the lirc.org website.
  * Old, non-functional kernel modules are removed.
  * The configuration data which was in several places including
    configure.ac and setup.data is now in the configs/ directory.

  * Support for sending data has been added to the lirc_client.h API.
    (ticket #35).
  * lirc now installs three separate libraries: lirc_private (internal),
    lirc (for external applications) and lirc_driver for drivers. Each
    library has a so-file to link to, a  header file for compilation
    and a pkg-config .pc file for build. Applications  and drivers have
    been refactored to use the libraries.
  * There are examples in the sources how to build new drivers. The
    documentation in this area is also improved (see the driver-api guide
    and doc/api-docs).
  * The doc/ directory has support for creating doxygen API documentation.

tools/mode2.c

/****************************************************************************
 ** mode2.c *****************************************************************
 ****************************************************************************
 *
 * mode2 - shows the pulse/space length of a remote button
 *
 * Copyright (C) 1998 Trent Piepho <xyzzy@u.washington.edu>
 * Copyright (C) 1998 Christoph Bartelmus <lirc@bartelmus.de>
 *
 */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#ifdef _CYGWIN_
#define __USE_LINUX_IOCTL_DEFS
#endif

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <limits.h>
#include <errno.h>
#include <syslog.h>
#include <time.h>
#include <pwd.h>

#include "lirc_private.h"

static char *opt_device = NULL;
static int opt_dmode = 0;
static int t_div = 500;
static int opt_gap = 10000;
static int opt_raw_access = 0;


static const char* const help =
"Usage: mode2 [options]\n"
"\t -d --device=device\tread from given device\n"
"\t -H --driver=driver\tuse given driver\n"
"\t -U --plugindir=path\tLoad plugins from path.\n"
"\t -k --keep-root\t\tkeep root privileges\n"
"\t -m --mode\t\tenable column display mode\n"
"\t -r --raw\t\taccess device directly\n"
"\t -g --gap=time\t\ttreat spaces longer than time as the gap\n"
"\t -s --scope=time\tenable 'scope like display with time us per char.\n"
"\t -h --help\t\tdisplay usage summary\n"
"\t -v --version\t\tdisplay version\n"
"\t -A --driver-options=key:value[|key:value...]\n"
"\t\t\t\tSet driver options\n";

static const struct option options[] = {
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'v'},
	{"device", required_argument, NULL, 'd'},
	{"driver", required_argument, NULL, 'H'},
	{"keep-root", no_argument, NULL, 'k'},
	{"mode", no_argument, NULL, 'm'},
	{"raw", no_argument, NULL, 'r'},
	{"gap", required_argument, NULL, 'g'},
	{"scope", required_argument, NULL, 's'},
	{"plugindir", required_argument, NULL, 'U'},
	{"driver-options", required_argument, NULL, 'A'},
	{0, 0, 0, 0}
};

const char* const MSG_NO_GETMODE =
	"Problems: this device is not a LIRC kernel device (it does not\n"
	"support LIRC_GET_REC_MODE ioctl). This is not necessarily a\n"
	"problem, but mode2 will not work.  If you are using the --raw\n"
	"option you might try using without it and select a driver\n"
	"instead. Otherwise, try using lircd + irw to view the decoded\n"
	"data - this might very well work even if mode2 doesn't.";


static void add_defaults(void)
{
	char level[4];
	snprintf(level, sizeof(level), "%d", lirc_log_defaultlevel());

	const char* const defaults[] = {
		"mode2:driver", 	"default",
		"mode2:lircdfile", 	LIRCD,
		"lircd:logfile", 	"syslog",
		"lircd:debug", 		level,
		"lircd:plugindir", 	PLUGINDIR,
		"lircd:configfile",  	LIRCDCFGFILE,
		(const char*)NULL, 	(const char*)NULL
	};
	options_add_defaults(defaults);
}


static void parse_options(int argc, char** argv)
{
	int c;
	static const char* const optstring= "hvd:H:mkrg:s:U:A:";
	char driver[64];

	strcpy(driver, "default");
	add_defaults();
	while ((c = getopt_long(argc, argv, optstring, options, NULL)) != -1)
	{
		switch (c) {
		case 'h':
			puts(help);
			exit(EXIT_SUCCESS);
		case 'H':
			strncpy(driver, optarg, sizeof(driver) - 1);
			break;
		case 'v':
			printf("%s %s\n", "mode2 ", VERSION);
			exit(EXIT_SUCCESS);
		case 'U':
			options_set_opt("lircd:plugindir", optarg);
			break;
		case 'k':
			unsetenv("SUDO_USER");
			break;
		case 'd':
			opt_device = optarg;
			break;
		case 's':
			opt_dmode = 2;
			t_div = atoi(optarg);
			break;
		case 'm':
			opt_dmode = 1;
			break;
		case 'r':
			opt_raw_access = 1;
			break;
		case 'g':
			opt_gap = atoi(optarg);
			break;
		case 'A':
			options_set_opt("lircd:driver-options", optarg);
			break;
		default:
			printf("Usage: mode2 [options]\n");
			exit(EXIT_FAILURE);
		}
	}
	if (optind < argc) {
		fputs("Too many arguments\n", stderr);
		exit(EXIT_FAILURE);
	}
	if (opt_raw_access && opt_device == NULL ) {
		fprintf(stderr, "The --raw option requires a --device\n");
		exit(EXIT_FAILURE);
	}
	if (hw_choose_driver(driver) != 0) {
		fprintf(stderr, "Driver `%s' not found.", driver);
		fputs(" (Missing -U/--plugins option?)\n", stderr);
		fputs("Available drivers:\n", stderr);
		hw_print_drivers(stderr);
		exit(EXIT_FAILURE);
	}
}

/** Open device using curr_driver->open_func() and curr_driver->init_func().*/
int open_device(int opt_raw_access, const char* device)
{
	struct stat s;
	__u32 mode;
	int fd;
	const char* opt;

	if (opt_raw_access) {
		fd = open(device, O_RDONLY);
		if (fd == -1) {
			perror("Error while opening device");
			exit(EXIT_FAILURE);
		};
		if ((fstat(fd, &s) != -1) && (S_ISFIFO(s.st_mode))) {
			/* can't do ioctls on a pipe */
		} else if ((fstat(fd, &s) != -1) && (!S_ISCHR(s.st_mode))) {
			fprintf(stderr, "%s is not a character device\n",
				device);
			fputs("Use the -d option to specify device\n",
			       stderr);
			close(fd);
			exit(EXIT_FAILURE);
		} else if (ioctl(fd, LIRC_GET_REC_MODE, &mode) == -1) {
			puts(MSG_NO_GETMODE);
			close(fd);
			exit(EXIT_FAILURE);
		}
	} else {
		curr_driver->open_func(device);
		opt = options_getstring("lircd:driver-options");
		if (opt != NULL)
			if (drv_handle_options(opt) != 0) {
				fprintf(stderr,
				"Cannot set driver (%s) options (%s)\n",
				curr_driver->name, opt);
				exit(EXIT_FAILURE);
			}
		if (!curr_driver->init_func || !curr_driver->init_func()) {
			fprintf(stderr, "Cannot initiate device %s\n",
			 	curr_driver->device);
			exit(EXIT_FAILURE);
		}
		printf("Using device: %s\n", curr_driver->device);
		fd = curr_driver->fd;	/* please compiler */
		mode = curr_driver->rec_mode;
		if (mode != LIRC_MODE_MODE2) {
			if (strcmp(curr_driver->name, "default") == 0) {
				puts("Please use the --raw option to access "
				       "the device directly instead through\n"
				       "the abstraction layer");
				exit(EXIT_FAILURE);
			} else if (mode != LIRC_MODE_LIRCCODE) {
				puts("Internal error: bad receive mode");
				exit(EXIT_FAILURE);
			}
		}
	}
	if (opt_device && strcmp(opt_device, LIRCD) == 0) {
		fputs("Refusing to connect to lircd socket\n", stderr);
		return EXIT_FAILURE;
	}
	printf("Using device: %s\n", curr_driver->device);
	return fd;
}


/** Define loglevel and open the log file. */
static void setup_log(void)
{
	char path[128];
	const char* loglevel;

	loglevel = getenv("LIRC_LOGLEVEL");
	if (loglevel == NULL) {
		loglevel = "LIRC_NOTICE";
	} else if (string2loglevel(loglevel) == LIRC_BADLEVEL) {
		fprintf(stderr, "Bad LIRC_LOGLEVEL: %s\n", loglevel);
		loglevel = "LIRC_NOTICE";
	}
	lirc_log_get_clientlog("mode2", path, sizeof(path));
	lirc_log_set_file(path);
	lirc_log_open("mode2", 1, string2loglevel(loglevel));
}


/** Get the codelength (bits per decoded value) for lirccode data. */
unsigned int get_codelength(int fd, int use_raw_access)
{
	unsigned int code_length;

	if (use_raw_access) {
		if (ioctl(fd, LIRC_GET_LENGTH, &code_length) == -1) {
			perror("Could not get code length");
			close(fd);
			exit(EXIT_FAILURE);
		}
	} else {
		code_length = curr_driver->code_length;
	}
	if (code_length > sizeof(ir_code) * CHAR_BIT) {
		fprintf(stderr, "Cannot handle %u bit codes\n",
			code_length);
		close(fd);
		exit(EXIT_FAILURE);
	}
	return code_length;
}


/** Print mode2 data as pulse/space durations, one per line. */
void print_mode2_data(unsigned int data)
{
	static int bitno = 1;

	switch (opt_dmode) {
	case 0:
		printf("%s %u\n", (
		       data & PULSE_BIT) ? "pulse" : "space",
		       (__u32) (data & PULSE_MASK));
		break;
	case 1: {
		/* print output like irrecord raw config file data */
		printf(" %8u", (__u32) data & PULSE_MASK);
		++bitno;
		if (data & PULSE_BIT) {
			if ((bitno & 1) == 0) {
				/* not in expected order */
				fputs("-pulse", stdout);
			}
		} else {
			if (bitno & 1) {
				/* not in expected order */
				fputs("-space", stdout);
			}
			if (((data & PULSE_MASK) > opt_gap) || (bitno >= 6)) {
				/* real long space or more
				   than 6 codes, start new line */
				puts("");
				if ((data & PULSE_MASK) > opt_gap)
					puts("");
				bitno = 0;
			}
		}
		break;
	}
	case 2:
		if ((data & PULSE_MASK) > opt_gap)
			fputs("_\n\n_", stdout);
		else
			printf("%.*s",
			       ((data & PULSE_MASK) + t_div/2) / t_div,
			       (data & PULSE_BIT) ?
					"------------" : "____________");
		break;
	}
	fflush(stdout);
}


/** Print lirccode data as a decoded value (an integer) per line. */
void print_lirccode_data(char* buffer, size_t count)
{
	int i;

	fputs("code: 0x", stdout);
	for (i = 0; i < count; i++) {
		printf("%02x", (unsigned char)buffer[i]);
	}
	puts("");
	fflush(stdout);
}


int main(int argc, char **argv)
{
	int fd;
	char buffer[sizeof(ir_code)];
	lirc_t data;
	__u32 mode;
	size_t count = sizeof(lirc_t);
	const char* new_user;
	int result;
	/**
	 * Was hard coded to 50000 but this is too long, the shortest gap in the
	 * supplied .conf files is 10826, the longest space defined for any one,
	 * zero or header is 7590
	 */
	__u32 code_length;

	hw_choose_driver(NULL);
	options_load(argc, argv, NULL, parse_options);
	setup_log();
	fd = open_device(opt_raw_access, opt_device);
	new_user = drop_sudo_root(setuid);
	if (strcmp("root", new_user) == 0)
		puts("Warning: Running as root.");
	else if (strlen(new_user) == 0)
		puts("Warning: Cannot change uid.");
	else
		printf("Running as regular user %s\n", new_user);
	mode = curr_driver->rec_mode;
	if (mode == LIRC_MODE_LIRCCODE) {
		code_length = get_codelength(fd, opt_raw_access);
		count = (code_length + CHAR_BIT - 1) / CHAR_BIT;
	}
	while (1) {
		if (opt_raw_access || mode != LIRC_MODE_MODE2) {
			result = read(fd,
				      (mode == LIRC_MODE_MODE2 ?
						(void *)&data : buffer),
				      count);
			if (result != count) {
				fputs("read() failed\n", stderr);
				break;
			}
			if (mode == LIRC_MODE_MODE2)
				print_mode2_data(data);
			else
				print_lirccode_data(buffer, count);
		} else {
			data = curr_driver->readdata(0);
			if (data == 0) {
				fputs("readdata() failed\n", stderr);
				break;
			}
			print_mode2_data(data);
		}
	}
	return (EXIT_SUCCESS);
}
